design/kopia.md

```diff
@@ -0,0 +1,461 @@
+during restart.
+
+## Repository Lifecycle
+
+When a `Repository` resource is created, the `RepositoryController` controller
+responds by creating a new Kopia repository using the equivalent of this
+command:
+
+```sh
+kopia repository create s3 \
+  --bucket=my-bucket \
+  --access-key=<redacted> \
+  --secret-access-key=<redacted> \
+  --create-only=true \
+  [--endpoint=https://foo.example.com \]
+  [--prefix=my-prefix \]
+  [--region=us-west-1]
+```
+
+> 📝 The examples in this section uses S3 for illustration purposes. Other
+> supported storage types include GCS And Azure Blob Storage.
+
+### Server-Side Setup
+
+Once the repository is ready, the controller will schedule a Kopia server
+instance to run in the `kanister` namespace, as a `Deployment` workload. The
+server is started using the equivalent of this command:
+
+```sh
+kopia server start --address=0.0.0.0:51515 \
+  --config-file=/run/kopia/repo.config \
+  --tls-cert-file=/run/kopia/tls.crt \
+  --tls-key-file=/run/kopia/tls.key \
+  --server-username=<redacted> \
+  --server-password=<redacted> \
+  --server-control-username=<redacted> \
+  --server-control-password=<redacted> \
+```
+
+The `/run/kopia/repo.config` configuration file is generated from the data found
+in the secret referenced by the repository's `spec.location.secretName` property.
+See the [Kopia documentation][3] and [GitHub source code][4] for more
+information on the configuration file format, and supported configuration.
+
+The `/run/kopia/tls.crt` and `/run/kopia/tls.key` files contain the TLS x509
+certificate and private key read from the secret referenced by the repository's
+`spec.server.tlsSecretName` property.
+
+The credentials for the `--server-username`, `--server-password`,
+`--server-control-username` and `--server-control-password` options are read
+from the secret referenced by the `spec.server.adminSecretName` property.
+
+### Client-Side Setup
+
+The Kopia server is fronted by a K8s `Service` resource. Data mover clients will
+connect to it using the equivalence of:
+
+```sh
+kopia repository connect server \
+  --url=https://<service-name>.kanister.svc.cluster.local \
+  --config-file=/run/kopia/repo.config \
+  --server-cert-fingerprint=<redacted> \
+  --password=<redacted>
+```
+
+The `<service-name>` is the Kopia server's `Service` resource name. By
+convention, it includes the repository's `targetedNamespace`. E.g., the
+`Service` fronting the Kopia server of a repository serving the
+`monitoring` namespace can be `repo-monitoring`. Then its FQDN is
+`repo-monitoring.kanister.svc.cluster.local`.
+
+The access username used to authenticate with the Kopia server is added to the
+`/run/kopia/repo.config` configuration file, following the
+[`repo.ClientOptions`][5] structure.
+
+The content of the configuration file, server certificate fingerprint, and
+and access password are exported to the remote data mover clients via the
+`KubeExec` functionality over TLS.
```

Comment:
Not sure if I completely understand this. Would we use `outputArtifacts` to pass this information to all other `BackupData`, `RestoreData`, etc. actions defined in BP?

Reply:
Yeah, this part still needs to be fleshed out. With `KubeTask`, we have more control over what goes into the container before it's started. With `KubeExec`, current thought is to `POST` these info to the `exec` subresource of the data mover pod via its `stdin`.